### PR TITLE
Switch flags to PFlag to remove StringArray

### PR DESCRIPTION
**Pull request by JoelSpeed**
## Description

This is a follow up to #484 to switch all flag loading to PFlag. Currently based on #484 so will need a rebase once that is merged. Only the last commit is new.

## Motivation and Context

Allows us to delete `StringArray`

## How Has This Been Tested?

This should be a fairly simple change, tests from #484 check PFlag behaviour matches the `StringArray` behaviour already

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [ ] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -42,7 +31,7 @@ func main() {
 	flagSet.String("tls-key-file", "", "path to private key file")
 	flagSet.String("redirect-url", "", "the OAuth Redirect URL. ie: \"https://internalapp.yourcompany.com/oauth2/callback\"")
 	flagSet.Bool("set-xauthrequest", false, "set X-Auth-Request-User and X-Auth-Request-Email response headers (useful in Nginx auth_request mode)")
-	flagSet.Var(&upstreams, "upstream", "the http url(s) of the upstream endpoint, file:// paths for static files or static://<status_code> for static response. Routing is based on the path")
+	flagSet.StringSlice("upstream", []string{}, "the http url(s) of the upstream endpoint, file:// paths for static files or static://<status_code> for static response. Routing is based on the path")
```
Comment (ploxiln) [inline main.go:34]: I think pflag`FlagSet.StringArray` has behavior closer to the previous internal `StringArray`, whereas this `FlagSet.StringSlice` splits values on comma (in addition to taking multiple flags/values).

**Comment by JoelSpeed** [comment]
@ploxiln You raise a really good point, this does slightly change the behaviour of the flags, which I hadn't realised.

Having just checked, the current options resolved (https://github.com/mreiferson/go-options/blob/6fa6a130bc8c7e1dc5c79adf48dfcdb1b55fe3ab/options.go#L157-L170) splits strings on commas in the current configuration anyway, so I don't think this will actually break the behaviour of the options overall, do you agree?

**Comment by ploxiln** [comment]
The coerce functions are called only if the value type passed does not match the option type, but oauth2-proxy prepares the value with type `[]string` before passing it. So with current master:

```
./oauth2-proxy ... --upstream="http://127.0.0.1:8080/first/,http://127.0.0.1:8082/second/"
[2020/04/18 11:41:38] [oauthproxy.go:216] mapping path "/first/,http://127.0.0.1:8082/second/" => upstream "http://127.0.0.1:8080/first/,http://127.0.0.1:8082/second/"
```

but maybe people would like the split-on-comma feature added anyway :)

**Comment by JoelSpeed** [comment]
Ahh yes I see, teaches me for scanning 😅

I had a quick look at most options that use these string arrays, I don't think any valid value for any of them would contain a comma anyway, they are mostly domains

I think I might keep this as is but add a note to the changelog to let users know this has changed

File: main.go (hunk 1 above)
Comment (steakunderscore) [reply main.go:34]: I think the old behaviour was closer to `StringArray` (based off this [description](https://github.com/spf13/pflag/issues/89)). But, at least for upstream where you can't have a comma in a url then using `StringSlice` would be better.

**Pull request merged**